### ISSUE TITLE
Update the parent ID of spans

### ISSUE THREAD

**Issue by felixbarny** [elastic/apm-server#3422]
This is a feature needed to improve the experience of profiler-inferred spans.

This aims to fix a problem outlined here: https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-profiling-plugin/README.md#with-non-implemented-workaround

Because inferred spans are created after-the-fact, a regular span cannot be a child of an inferred span.

Hence, we need a way to modify the parent id of spans after they are created.

## Intake API changes
We'd need a new event type, either a specific one to just change the parent id or a generic one which allows updating arbitrary properties.

Just some ideas on how to tackle this:
## Approach 1: update the ID in Elasticsearch

Use Elasticsearch's update API to update the parent id

Challenges:

A span can be in the following states:
- Waiting in the agent's internal queue 
- On the wire in-between the agent and APM Server
- Waiting in APM Server's queue to be transmitted to Elasticsearch
- Indexed in Elasticsearch

There's even a chance that an ID update requests arrives at Elasticsearch before the corresponding span arrives. That can happen if the update request is handled by another API Server, for example.

In summary, we can't be sure whether a span is already in Elasticsearch when we want to update the ID. 

## Approach 2: index update document alongside span document

Instead of actually changing the parent id we could just index a separate document that sits alongside a span that contains the updated properties.

Challenges: The logic in the UI that constructs the waterfall tree has to incorporate this.

**Comment by graphaelli** [comment]
I think the timing issues in approach 1 make it a non-starter.  Approach 2 seems doable but would prefer a prototype to confirm before settling on this.  Another option discussed and discarded to simply aligning inferred spans to the nearest parent by time, as async spans still would not show up correctly.  @elastic/apm-server any other ideas on this?

**Comment by axw** [comment]
Approach 1 could be made to work, but would require some significant changes:
 - Specify the _id for span docs (e.g. `trace.id` + `span.id`)
 - Write these "change" docs to a special change-queue index
 - Have something (APM Server? Kibana?) process this queue, performing _bulk "update" operations, followed by deleting the change docs. If the original doc doesn't exist yet, retry the change later

This could be a useful approach if we want to enable updating things other than IDs – I don't think we'd want to have multiple docs for changing context fields as it would mess up aggregations. It also has the nice outcome of sticking with one doc per span.

OTOH, this approach has some pretty significant drawbacks:
 - Much more complex
 - Requires additional writes and reads to process the change queue
 - Explicitly setting _id on docs has a performance impact due to the need for ES to check for an existing doc before writing
 - Probably more lag until updates are applied?

Of these two options, I'd be inclined to go with approach 2: introduce a "reparent" event, which would have the singular purpose of changing another event's parent ID. Multiple reparent events for the same event would have non-deterministic behaviour.

---

There is a more radical third option, which would be to introduce a field on spans to set its _children_. This would take us in the direction of [causal ordering](https://en.wikipedia.org/wiki/Happened-before), reinterpreting "B's parent is A" to mean "B happens-after A", and "B's child is C" to mean "C happens-after B".

Example:

- Transaction T1
- S1's parent = T1
- S3a's parent = S1
- S3b's parent = S1

Then we report an inferred span, S2, whose parent is S1 and which should be the new parent of S3a and S3b.

- S1's parent = T1 => S1 happens-after T1
- S3a's parent = S1 => S3a happens-after S1
- S3b's parent = S1 => S3b happens-after S1
- S2's parent = S1 => S2 happens-after S1
- S2'd children are [S3a, S3b] => [S3a, S3b] happens-after S2

```dot
digraph graphname {
  T1 -> S1;

  S1 -> S3a [style=dotted];
  S1 -> S3b [style=dotted];
  S1 -> S2;

  S2 -> S3a;
  S2 -> S3b;
}
```

![foo](https://user-images.githubusercontent.com/843579/77243795-104b3c80-6c49-11ea-95b0-30ac9058c091.png)

When creating the waterfall visualisation, the UI would remove direct edges from S1 to [S3a, S3b] via [Arborescence](https://en.wikipedia.org/wiki/Arborescence_(graph_theory)).

**Comment by graphaelli** [comment]
@axw Can you elaborate on why option 3 is radical and isn't the obvious choice?  Assuming an agent has the information available to fill in the children it seems that would be of a similar complexity (from the UI perspective) to incorporating a 2nd index of.information  

Also, since there is no folding currently in the UI is the removal of direct edges from the former parent relevant in terms of generating the waterfall?

**Comment by axw** [comment]
@graphaelli mostly because "parent" would no longer mean "parent". If we called it "predecessor" (and called "child", "successor"), then I think it would be the way to go. We could also introduce new field names.

> Also, since there is no folding currently in the UI is the removal of direct edges from the former parent relevant in terms of generating the waterfall?

It's only relevant when we want to visualise the causality/introduce folding. Doesn't this hold true already? Is there currently a practical impact due to having incorrect parent IDs on spans?

**Comment by graphaelli** [comment]
@axw Understood.  I'd be comfortable with loosening the definitions of parent/child to accommodate this but agree (per our oob discussion) to put it to a proposal for wider input.

@felixbarny would you chime in here regarding feasibility of inferred spans reporting their children, in addition to the current parent information?  Specifying successors may be a useful concept either even if the can't be used with the current inferred span implementation but we'll need to prioritize the issue here above that.

**Comment by felixbarny** [comment]
It's probably feasible to do but it's more complicated than I originally anticipated. The complications are not specific to the transitive reduction approach. I didn't think about cases where there is not exactly one inferred span as a child of a regular span.

That is because currently, the Java agent only stores the latest activation as the parent of a CallTree node. It needs to be a stack of child activations for a CallTree node instead. This would translate well to reporting the child spans of an inferred span.

The downside is that this would increase the allocations necessary in the Java agent. Previously, consecutive activations without an inferred span in-between could be ignored. But in order to fix the parent/child relationships, there's no way around it and I think it's still feasible from a performance standpoint.

I can only definitely confirm that this works out after I have implemented it in the Java agent. But I'm optimistic that it should work out.

How should the intake API look like? Would I just send a string array of child ids or does it need to be an array of traceId, transactionId, spanId tuples? I'd vote for just sending the spanIds and imply that the transactions and traceId are the same.

**Comment by axw** [comment]
> How should the intake API look like? Would I just send a string array of child ids or does it need to be an array of traceId, transactionId, spanId tuples? I'd vote for just sending the spanIds and imply that the transactions and traceId are the same.

I think we can send just the child/successor IDs, and imply the same trace ID.

Span IDs need to be unique within a trace anyway, so no need to imply a transaction ID. Also, that means we can maintain the idea that transactions are just special spans, and allow transaction IDs to be specified in the child IDs.

So, something like:

```json
"span": {
    "id": "span1",
    "trace_id": "traceid1",
    "transaction_id": "transactionid1",
    "parent_id": "transactionid1",
    "child_ids": ["span2", "span3", "transactionid2"]
}
```

(and perhaps `s/child_ids/successor_ids/`)

**Comment by felixbarny** [comment]
I can confirm that the `child_ids`/`successor_ids` approach works out with the Java agent: https://github.com/elastic/apm-agent-java/pull/1117

**Comment by graphaelli** [comment]
That's great news.  I will pull together an issue summarizing the intake and data model changes proposed so we can get input from UI and other agents before we proceed.
